### PR TITLE
refactor: data request mappers

### PR DESCRIPTION
**Pull request by Judge40**
Refactor the DMS DTO mapper and their usage to allow a more simplified
way to add additional data types.

Create a parameterised enum which can be selected by TCS DTO class and
provides access to the appropriate schema, table and mapper to convert
it in to a DMS DTO.
Refactor DmsRecordAssembler to make use of the new enum to generically
retrieve and use the correct mapper class.
Create DmsMapper and refactor existing mappers to provide the ability to
call any mapper using the same method call.

TIS21-2997
TIS21-2999

### CODE REVIEW

File: src/main/resources/config/application-local.yml [hunk 1]
```diff
@@ -39,9 +39,9 @@ enable.es.search: false
...
   cron:
-    recordResendingJob: ${APPLICATION_CRON_RECORDRESENDINGJOB:-}
+    recordResendingJob: ${APPLICATION_CRON_RECORDRESENDINGJOB:0 * * * * *}
   aws:
     kinesis:
-      streamName: ${AWS_KINESIS_STREAM_NAME:}
+      streamName: ${AWS_KINESIS_STREAM_NAME:tis21-2997_test_stream}
     sqs:
-      queueUrl: ${AWS_SQS_QUEUE_URL:}
+      queueUrl: ${AWS_SQS_QUEUE_URL:https://sqs.eu-west-2.amazonaws.com/430723991443/tis21-2997_test_queue}
```
Comment (JosephKelly) [inline src/main/resources/config/application-local.yml:47]: Did you mean to commit these?

**Review by JosephKelly — COMMENTED** [review]
Looks like a good way of minimising any change in paradigm.
Like I said for trainee-sync I'm probably a bit "over keen" on using enums so that seems like a good way of encapsulating the DMS info.

I haven't worked out how to reduce the "appearence" of if/else blocks by supplying or declaring a Function<T,T> if that makes sense.

**Comment by hihilary** [comment]
I guess (PlacementDetailsMapper, PlacementDetailsDmsDto), (ProgrammeMapper, ProgrammeDmsDto) can be removed too?

**Review by JosephKelly — COMMENTED** [review]
By using the TCS & Reference service DTOs, we'd end up including nested objects, e.g. an `SiteDTO.organizationType`.
It might not break downstream functionality but wanted to make sure it was a conscious choice to stop matching the DMS produced data records.

**Comment by Judge40** [comment]
> I guess (PlacementDetailsMapper, PlacementDetailsDmsDto), (ProgrammeMapper, ProgrammeDmsDto) can be removed too?

Unfortunately not, the TCS DTOs have nested objects included so we still need DMS DTOs to simplify/map back to something representative of the DB table.

**Comment by Judge40** [comment]
> By using the TCS & Reference service DTOs, we'd end up including nested objects, e.g. an `SiteDTO.organizationType`. It might not break downstream functionality but wanted to make sure it was a conscious choice to stop matching the DMS produced data records.

Maybe not transparent enough, but the DMS DTOs that were removed were those that mirrored the TCS DTOs with just an ID type change (Long -> String) which has been handled elsewhere.
Anything with nested DTOs will still need a DMS DTO/Mapper, there is definitely an argument to be made that upstream changes to the TCS DTOs could cause issues though 🤔

**Comment by hihilary** [comment]
> ProgrammeMapper

I noticed for the moment, `ProgrammeMapper` and `PlacementDetailsMapper` don't use a nested mapper for nested objects, so I assume they're kept for future changes.

**Comment by hihilary** [comment]
> > I guess (PlacementDetailsMapper, PlacementDetailsDmsDto), (ProgrammeMapper, ProgrammeDmsDto) can be removed too?
> 
> Unfortunately not, the TCS DTOs have nested objects included so we still need DMS DTOs to simplify/map back to something representative of the DB table.

I noticed for the moment, `ProgrammeMapper` and `PlacementDetailsMapper` don't use a nested mapper for nested objects, so I assume they're kept for future changes.

File: src/main/java/uk/nhs/tis/sync/mapper/TrustMapper.java [hunk 2]
```diff
@@ -1,11 +0,0 @@
-package uk.nhs.tis.sync.mapper;
-
-import com.transformuk.hee.tis.reference.api.dto.TrustDTO;
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/tis/sync/mapper/TrustMapper.java:3]: TrustDTO has a nested `com.transformuk.hee.tis.reference.api.dto.OrganizationTypeDto`

File: src/main/java/uk/nhs/tis/sync/dto/TrustDmsDto.java [hunk 3]
```diff
@@ -1,20 +0,0 @@
-package uk.nhs.tis.sync.dto;
-
-import com.fasterxml.jackson.annotation.JsonInclude;
-import lombok.Data;
-import lombok.NoArgsConstructor;
-
-@NoArgsConstructor
-@JsonInclude(JsonInclude.Include.NON_NULL)
-@Data
-public class TrustDmsDto {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/tis/sync/dto/TrustDmsDto.java:10]: As per mapper, the source has a nested DTO. There are 779 "CURRENT" records with a value.

File: src/main/java/uk/nhs/tis/sync/dto/SiteDmsDto.java [hunk 4]
```diff
@@ -1,26 +0,0 @@
-package uk.nhs.tis.sync.dto;
-
-import com.fasterxml.jackson.annotation.JsonInclude;
-import com.fasterxml.jackson.annotation.JsonInclude.Include;
-import lombok.Data;
-
-@Data
-@JsonInclude(Include.NON_NULL)
-public class SiteDmsDto {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/tis/sync/dto/SiteDmsDto.java:9]: As per mapper, the source has a nested DTO. There are only 2 records _without_ an endDate in the past.

File: src/main/java/uk/nhs/tis/sync/mapper/DmsMapper.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+   *
+   * @param object The object to convert.
+   * @return The converted object, or null if the object was not of the correct class.
+   */
+  default O objectToDmsDto(Object object) {
+    try {
+      return toDmsDto((I) object);
+    } catch (ClassCastException e) {
+      return null;
+    }
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/tis/sync/mapper/DmsMapper.java:30]: This seems ok and consistent with the existing expectations.

As getting to this point looks like really unexpected behaviour, logging an error would be a useful distinction.  It seems best pushed out to the caller. What do you think?

**Review by JosephKelly — CHANGES_REQUESTED** [review]
> > By using the TCS & Reference service DTOs, we'd end up including nested objects, e.g. an SiteDTO.organizationType. It might not break downstream functionality but wanted to make sure it was a conscious choice to stop matching the DMS produced data records.
>
> Anything with nested DTOs will still need a DMS DTO/Mapper

I've marked where I see nested DTOs related to 2 Mappers deleted in the PR.

File: src/main/java/uk/nhs/tis/sync/dto/TrustDmsDto.java (hunk 3 above)
Comment (Judge40) [reply src/main/java/uk/nhs/tis/sync/dto/TrustDmsDto.java:10]: As with SiteDTO, the version of reference-api used doesn't include that nested DTO.

File: src/main/java/uk/nhs/tis/sync/mapper/TrustMapper.java (hunk 2 above)
Comment (Judge40) [reply src/main/java/uk/nhs/tis/sync/mapper/TrustMapper.java:3]: See other comments regarding reference-api version

File: src/main/java/uk/nhs/tis/sync/mapper/DmsMapper.java (hunk 5 above)
Comment (Judge40) [reply src/main/java/uk/nhs/tis/sync/mapper/DmsMapper.java:30]: A log entry makes sense, are you thinking we should let the exception be thrown though? I guess its a scenario that should never happen unless someone messes up the enum params, so maybe letting the ClassCastException be thrown is okay.

Comment (Judge40) [reply src/main/java/uk/nhs/tis/sync/mapper/DmsMapper.java:30]: I'm never keen on declaring unchecked exceptions as thrown. Maybe just take out the catch and add a note in the javadoc?

Comment (JosephKelly) [reply src/main/java/uk/nhs/tis/sync/mapper/DmsMapper.java:30]: I saw this as one of those "while you're there"*, things to consider.  Leaving as is, if only for now, is also ok for me.

*(looking at other PR comments)

**Comment by Judge40** [comment]
The programme membership mapper was added and merged before this refactoring has been merged, I've migrated it to use this refactored mapper after merging main in to this branch.
In addition I've rewritten the associated test to compare individual attributes as the failure message was difficult to read when comparing equality between the objects directly.

**Review by hihilary — APPROVED** [review]

**Comment by ReubenRobertsHEE** [comment]
If possible, would you consider not pushing this to prod before taking a look at https://github.com/Health-Education-England/TIS-SYNC/pull/181? This fixes a bug with the programme membership DTO, though it doesn't play very nice with the DmsRecordAssembler() :(

**Pull request merged**